From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wosnxn123 <wosnxn123@users.noreply.github.com>
Date: Mon, 18 Aug 2026 12:00:00 +0800
Subject: [PATCH] Luminol: Configurable region format framework

Ported from Winds-Studio/Leaf 0107-Luminol-Configurable-region-format-
framework.patch at a05f8902de772298bbdf28142eef6cef003ea5c7 (Leaf ver/26.2),
originally LuminolMC/Luminol (region format framework, GPL-3.0-only) and
xymb-endcrystalme/Abomination (Linear V2, GPL-3.0-only).

Enables LINEAR_V2 and B_LINEAR region file formats (zstd+LZ4, ~50% disk
savings) alongside vanilla MCA, selected via canvas-server.yml
region-format.* (see GlobalConfiguration.RegionFormat).

Adaptations for Canvas: abomination.* and me.earthme.luminol.* packages
relocated to io.canvasmc.canvas.regionformat; Leaf's RegionFormatConfig
(TOML) replaced by GlobalConfiguration.RegionFormat with initFormat()
hooked into postLoad; oversized-chunk hunk re-anchored onto Canvas's
rebranded log message; all pre-image blobs regenerated against the
Canvas applied tree.

diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/io/ChunkSystemRegionFileStorage.java b/ca/spottedleaf/moonrise/patches/chunk_system/io/ChunkSystemRegionFileStorage.java
index a814512fcfb85312474ae2c2c21443843bf57831..cf56577c5a8f03a9113393bc995c63847b23da67 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/io/ChunkSystemRegionFileStorage.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/io/ChunkSystemRegionFileStorage.java
@@ -8,9 +8,9 @@ public interface ChunkSystemRegionFileStorage {
 
     public boolean moonrise$doesRegionFileNotExistNoIO(final int chunkX, final int chunkZ);
 
-    public RegionFile moonrise$getRegionFileIfLoaded(final int chunkX, final int chunkZ);
+    public io.canvasmc.canvas.regionformat.IRegionFile moonrise$getRegionFileIfLoaded(final int chunkX, final int chunkZ); // Luminol - Configurable region file format
 
-    public RegionFile moonrise$getRegionFileIfExists(final int chunkX, final int chunkZ) throws IOException;
+    public io.canvasmc.canvas.regionformat.IRegionFile moonrise$getRegionFileIfExists(final int chunkX, final int chunkZ) throws IOException; // Luminol - Configurable region file format
 
     public MoonriseRegionFileIO.RegionDataController.WriteData moonrise$startWrite(
             final int chunkX, final int chunkZ, final CompoundTag compound
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java b/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
index 3bb9b58ee97464687e348e23b99159226415c267..9231c8a1fbfb6c531755472e406380148c03acf8 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
@@ -1273,7 +1273,7 @@ public final class MoonriseRegionFileIO {
                 this.regionDataController.finishWrite(this.chunkX, this.chunkZ, writeData);
                 // Paper start - flush regionfiles on save
                 if (this.world.paperConfig().chunks.flushRegionsOnSave) {
-                    final RegionFile regionFile = this.regionDataController.getCache().moonrise$getRegionFileIfLoaded(this.chunkX, this.chunkZ);
+                    final io.canvasmc.canvas.regionformat.IRegionFile regionFile = this.regionDataController.getCache().moonrise$getRegionFileIfLoaded(this.chunkX, this.chunkZ); // Luminol - Add configurable region file
                     if (regionFile != null) {
                         regionFile.flush();
                     } // else: evicted from cache, which should have called flush
@@ -1489,7 +1489,7 @@ public final class MoonriseRegionFileIO {
 
         public static interface IORunnable {
 
-            public void run(final RegionFile regionFile) throws IOException;
+            public void run(final io.canvasmc.canvas.regionformat.IRegionFile regionFile) throws IOException; // Luminol - Configurable region file format
 
         }
     }
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/storage/ChunkSystemChunkBuffer.java b/ca/spottedleaf/moonrise/patches/chunk_system/storage/ChunkSystemChunkBuffer.java
index 51c126735ace8fdde89ad97b5cab62f244212db0..e84729996f5f109d98544591874d6ae04af0cedf 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/storage/ChunkSystemChunkBuffer.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/storage/ChunkSystemChunkBuffer.java
@@ -8,5 +8,5 @@ public interface ChunkSystemChunkBuffer {
 
     public void moonrise$setWriteOnClose(final boolean value);
 
-    public void moonrise$write(final RegionFile regionFile) throws IOException;
+    public void moonrise$write(final io.canvasmc.canvas.regionformat.IRegionFile regionFile) throws IOException; // Luminol - Configurable region file format
 }
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 5abf060a4fcc027dee9eb076133576e4de9e9a57..d5e9975a36cf405ca3c1012a193d4b733c625293 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -970,10 +970,10 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         if (flush) {
             for (ServerLevel level : this.getAllLevels()) {
                 String storageName = level.getChunkSource().chunkMap.getStorageName();
-                LOGGER.info("ThreadedAnvilChunkStorage ({}): All chunks are saved", LEGACY_WORLD_NAMES_FOR_REALMS_LOG.getOrDefault(storageName, storageName));
+                LOGGER.info("ThreadedChunkStorage ({}): All chunks are saved", LEGACY_WORLD_NAMES_FOR_REALMS_LOG.getOrDefault(storageName, storageName)); // Luminol - configurable region format
             }
 
-            LOGGER.info("ThreadedAnvilChunkStorage: All dimensions are saved");
+            LOGGER.info("ThreadedChunkStorage: All dimensions are saved"); // Luminol - configurable region format
         }
 
         return result;
diff --git a/net/minecraft/util/worldupdate/FileToUpgrade.java b/net/minecraft/util/worldupdate/FileToUpgrade.java
index a7f2cfa9277c898038f6b9e0a3401db51012b64c..4cf59e5b55a08c81f9904e6b29d02142cb8c39bf 100644
--- a/net/minecraft/util/worldupdate/FileToUpgrade.java
+++ b/net/minecraft/util/worldupdate/FileToUpgrade.java
@@ -4,5 +4,5 @@ import java.util.List;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.chunk.storage.RegionFile;
 
-public record FileToUpgrade(RegionFile file, List<ChunkPos> chunksToUpgrade) {
+public record FileToUpgrade(io.canvasmc.canvas.regionformat.IRegionFile file, List<ChunkPos> chunksToUpgrade) { // Luminol - Configurable region file format
 }
diff --git a/net/minecraft/util/worldupdate/RegionStorageUpgrader.java b/net/minecraft/util/worldupdate/RegionStorageUpgrader.java
index b228cd6c54e8c8998923ab332882d99092769c67..060a05c24252594741273ec8b1ad422bc5d273d3 100644
--- a/net/minecraft/util/worldupdate/RegionStorageUpgrader.java
+++ b/net/minecraft/util/worldupdate/RegionStorageUpgrader.java
@@ -36,7 +36,7 @@ import org.slf4j.Logger;
 public class RegionStorageUpgrader {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final String NEW_DIRECTORY_PREFIX = "new_";
-    private static final Pattern REGEX = Pattern.compile("^r\\.(-?[0-9]+)\\.(-?[0-9]+)\\.mca$");
+    private static final Pattern REGEX = Pattern.compile("^r\\.(-?[0-9]+)\\.(-?[0-9]+)\\." + io.canvasmc.canvas.GlobalConfiguration.get().regionFormat.format.getExtensionName() + "$"); // Luminol - Configurable region file format
     private final DataFixer dataFixer;
     private final UpgradeProgress upgradeProgress;
     private final String type;
@@ -176,7 +176,8 @@ public class RegionStorageUpgrader {
                 int zOffset = Integer.parseInt(regex.group(2)) << 5;
                 List<ChunkPos> chunkPositions = Lists.newArrayList();
 
-                try (RegionFile regionSource = new RegionFile(info, regionFile.toPath(), regionFolder, true)) {
+                var regionFileInfo = new io.canvasmc.canvas.regionformat.RegionCreatorInfo(info, regionFile.toPath(), regionFolder, true); // Luminol - Configurable region file format
+                try (io.canvasmc.canvas.regionformat.IRegionFile regionSource = io.canvasmc.canvas.GlobalConfiguration.get().regionFormat.format.getCreator().create(regionFileInfo)) { // Luminol - Configurable region file format
                     for (int x = 0; x < 32; x++) {
                         for (int z = 0; z < 32; z++) {
                             ChunkPos pos = new ChunkPos(x + xOffset, z + zOffset);
@@ -253,7 +254,7 @@ public class RegionStorageUpgrader {
         return storage.upgradeChunkTag(chunkTag, this.defaultVersion, this.dataFixContextTag, targetVersion);
     }
 
-    private void onFileFinished(final RegionFile regionFile) {
+    private void onFileFinished(final io.canvasmc.canvas.regionformat.IRegionFile regionFile) { // Luminol - Configurable region file format
         if (this.recreateRegionFiles) {
             if (this.previousWriteFuture != null) {
                 this.previousWriteFuture.join();
diff --git a/net/minecraft/world/level/chunk/storage/RegionFile.java b/net/minecraft/world/level/chunk/storage/RegionFile.java
index 1616ab8a89602a3cdc00a9b8608e42f316356273..51d282a7a2a9eadc65b70f23481a6bdf292048b1 100644
--- a/net/minecraft/world/level/chunk/storage/RegionFile.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFile.java
@@ -22,7 +22,7 @@ import net.minecraft.world.level.ChunkPos;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public class RegionFile implements AutoCloseable, ca.spottedleaf.moonrise.patches.chunk_system.storage.ChunkSystemRegionFile { // Paper - rewrite chunk system
+public class RegionFile implements AutoCloseable, ca.spottedleaf.moonrise.patches.chunk_system.storage.ChunkSystemRegionFile, io.canvasmc.canvas.regionformat.IRegionFile { // Paper - rewrite chunk system // Luminol - Configurable region file format
     private static final Logger LOGGER = LogUtils.getLogger();
     public static final int MAX_CHUNK_SIZE = 500 * 1024 * 1024; // Paper - don't write garbage data to disk if writing serialization fails
     private static final int SECTOR_BYTES = 4096;
@@ -907,7 +907,7 @@ public class RegionFile implements AutoCloseable, ca.spottedleaf.moonrise.patche
         }
 
         @Override
-        public final void moonrise$write(final RegionFile regionFile) throws IOException {
+        public final void moonrise$write(final io.canvasmc.canvas.regionformat.IRegionFile regionFile) throws IOException { // Luminol - Configurable region file format
             regionFile.write(this.pos, ByteBuffer.wrap(this.buf, 0, this.count));
         }
         // Paper end - rewrite chunk system
diff --git a/net/minecraft/world/level/chunk/storage/RegionFileStorage.java b/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
index 598facc98fc34e91eef77b1bdf86ddbe5ff86338..af2d80fdc6f0be728a8a97f7e814f1033df7a055 100644
--- a/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
@@ -19,7 +19,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     private static final org.slf4j.Logger LOGGER = com.mojang.logging.LogUtils.getLogger(); // Paper
     public static final String ANVIL_EXTENSION = ".mca";
     private static final int MAX_CACHE_SIZE = 256;
-    private final Long2ObjectLinkedOpenHashMap<RegionFile> regionCache = new Long2ObjectLinkedOpenHashMap<>();
+    private final Long2ObjectLinkedOpenHashMap<io.canvasmc.canvas.regionformat.IRegionFile> regionCache = new Long2ObjectLinkedOpenHashMap<>(); // Luminol - Configurable region file format
     private final RegionStorageInfo info;
     private final Path folder;
     private final boolean sync;
@@ -30,7 +30,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     @Nullable
     public static ChunkPos getRegionFileCoordinates(Path file) {
         String fileName = file.getFileName().toString();
-        if (!fileName.startsWith("r.") || !fileName.endsWith(".mca")) {
+        if (!fileName.startsWith("r.") || !fileName.endsWith(getExtensionName())) { // Luminol - Configurable region file format
             return null;
         }
 
@@ -55,8 +55,26 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     private static final int MAX_NON_EXISTING_CACHE = 1024 * 4;
     private final it.unimi.dsi.fastutil.longs.LongLinkedOpenHashSet nonExistingRegionFiles = new it.unimi.dsi.fastutil.longs.LongLinkedOpenHashSet();
     private static String getRegionFileName(final int chunkX, final int chunkZ) {
-        return "r." + (chunkX >> REGION_SHIFT) + "." + (chunkZ >> REGION_SHIFT) + ".mca";
+        return "r." + (chunkX >> REGION_SHIFT) + "." + (chunkZ >> REGION_SHIFT) + getExtensionName(); // Luminol - Configurable region file format
     }
+    // Luminol start - Configurable region file format
+    public static io.canvasmc.canvas.regionformat.IRegionFile createNew(final RegionStorageInfo info, final Path filePath, final Path folder, final boolean sync) throws IOException{
+        final io.canvasmc.canvas.regionformat.EnumRegionFormat regionFormat = io.canvasmc.canvas.GlobalConfiguration.get().regionFormat.format;
+        final String fullFileName = filePath.getFileName().toString();
+        final String[] fullNameSplit = fullFileName.split("\\.");
+        final String extensionName = fullNameSplit[fullNameSplit.length - 1];
+
+        if (!regionFormat.getExtensionName().equalsIgnoreCase(extensionName)) {
+            throw new IOException("Invalid region file format: " + extensionName + " expected " + regionFormat.getExtensionName());
+        }
+
+        return regionFormat.getCreator().create(new io.canvasmc.canvas.regionformat.RegionCreatorInfo(info, filePath, folder, sync));
+    }
+
+    public static String getExtensionName() {
+        return "." + io.canvasmc.canvas.GlobalConfiguration.get().regionFormat.format.getExtensionName();
+    }
+    // Luminol end - Configurable region file format
 
     private boolean doesRegionFilePossiblyExist(final long position) {
         synchronized (this.nonExistingRegionFiles) {
@@ -90,15 +108,15 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     }
 
     @Override
-    public synchronized final RegionFile moonrise$getRegionFileIfLoaded(final int chunkX, final int chunkZ) {
+    public synchronized final io.canvasmc.canvas.regionformat.IRegionFile moonrise$getRegionFileIfLoaded(final int chunkX, final int chunkZ) { // Luminol - Configurable region file format
         return this.regionCache.getAndMoveToFirst(ChunkPos.pack(chunkX >> REGION_SHIFT, chunkZ >> REGION_SHIFT));
     }
 
     @Override
-    public synchronized final RegionFile moonrise$getRegionFileIfExists(final int chunkX, final int chunkZ) throws IOException {
+    public synchronized final io.canvasmc.canvas.regionformat.IRegionFile moonrise$getRegionFileIfExists(final int chunkX, final int chunkZ) throws IOException { // Luminol - Configurable region file format
         final long key = ChunkPos.pack(chunkX >> REGION_SHIFT, chunkZ >> REGION_SHIFT);
 
-        RegionFile ret = this.regionCache.getAndMoveToFirst(key);
+        io.canvasmc.canvas.regionformat.IRegionFile ret = this.regionCache.getAndMoveToFirst(key); // Luminol - Configurable region file format
         if (ret != null) {
             return ret;
         }
@@ -124,7 +142,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
         FileUtil.createDirectoriesSafe(this.folder);
 
-        ret = new RegionFile(this.info, regionPath, this.folder, this.sync);
+        ret = this.createNew(this.info, regionPath, this.folder, this.sync); // Luminol - Configurable region file format
 
         this.regionCache.putAndMoveToFirst(key, ret);
 
@@ -143,7 +161,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
         }
 
         final ChunkPos pos = new ChunkPos(chunkX, chunkZ);
-        final RegionFile regionFile = this.getRegionFile(pos);
+        final io.canvasmc.canvas.regionformat.IRegionFile regionFile = this.getRegionFile(pos); // Luminol - Configurable region file format
 
         // note: not required to keep regionfile loaded after this call, as the write param takes a regionfile as input
         // (and, the regionfile parameter is unused for writing until the write call)
@@ -177,7 +195,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     ) throws IOException {
         final ChunkPos pos = new ChunkPos(chunkX, chunkZ);
         if (writeData.result() == ca.spottedleaf.moonrise.patches.chunk_system.io.MoonriseRegionFileIO.RegionDataController.WriteData.WriteResult.DELETE) {
-            final RegionFile regionFile = this.moonrise$getRegionFileIfExists(chunkX, chunkZ);
+            final io.canvasmc.canvas.regionformat.IRegionFile regionFile = this.moonrise$getRegionFileIfExists(chunkX, chunkZ); // Luminol - Configurable region file format
             if (regionFile != null) {
                 regionFile.clear(pos);
             } // else: didn't exist
@@ -192,7 +210,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     public final ca.spottedleaf.moonrise.patches.chunk_system.io.MoonriseRegionFileIO.RegionDataController.ReadData moonrise$readData(
         final int chunkX, final int chunkZ
     ) throws IOException {
-        final RegionFile regionFile = this.moonrise$getRegionFileIfExists(chunkX, chunkZ);
+        final io.canvasmc.canvas.regionformat.IRegionFile regionFile = this.moonrise$getRegionFileIfExists(chunkX, chunkZ); // Luminol - Configurable region file format
 
         final DataInputStream input = regionFile == null ? null : regionFile.getChunkDataInputStream(new ChunkPos(chunkX, chunkZ));
 
@@ -237,7 +255,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
             final ChunkPos pos = new ChunkPos(chunkX, chunkZ);
             final ChunkPos headerChunkPos = SerializableChunkData.getChunkCoordinate(ret);
-            final RegionFile regionFile = this.getRegionFile(pos);
+            final io.canvasmc.canvas.regionformat.IRegionFile regionFile = this.getRegionFile(pos); // Luminol - Configurable region file format
 
             if (regionFile.getRecalculateCount() != readData.recalculateCount()) {
                 return null;
@@ -261,7 +279,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     }
     // Paper end - rewrite chunk system
     // Paper start - rewrite chunk system
-    public RegionFile getRegionFile(ChunkPos pos) throws IOException {
+    public io.canvasmc.canvas.regionformat.IRegionFile getRegionFile(ChunkPos pos) throws IOException { // Luminol - Configurable region file format
         return this.getRegionFile(pos, false);
     }
     // Paper end - rewrite chunk system
@@ -273,7 +291,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
         this.isChunkData = info.dfuType()[0] == net.minecraft.util.datafix.DataFixTypes.CHUNK; // Paper - recalculate region file headers
     }
 
-    @org.jetbrains.annotations.Contract("_, false -> !null") private @Nullable RegionFile getRegionFile(final ChunkPos pos, boolean existingOnly) throws IOException { // CraftBukkit
+    @org.jetbrains.annotations.Contract("_, false -> !null") private io.canvasmc.canvas.regionformat.@Nullable IRegionFile getRegionFile(final ChunkPos pos, boolean existingOnly) throws IOException { // CraftBukkit // Luminol - Configurable region file format
         // Paper start - rewrite chunk system
         if (existingOnly) {
             return this.moonrise$getRegionFileIfExists(pos.x(), pos.z());
@@ -281,7 +299,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
         synchronized (this) {
             final long key = ChunkPos.pack(pos.x() >> REGION_SHIFT, pos.z() >> REGION_SHIFT);
 
-            RegionFile ret = this.regionCache.getAndMoveToFirst(key);
+            io.canvasmc.canvas.regionformat.IRegionFile ret = this.regionCache.getAndMoveToFirst(key); // Luminol - Configurable region file format
             if (ret != null) {
                 return ret;
             }
@@ -298,7 +316,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
             FileUtil.createDirectoriesSafe(this.folder);
 
-            ret = new RegionFile(this.info, regionPath, this.folder, this.sync);
+            ret = this.createNew(this.info, regionPath, this.folder, this.sync); // Luminol - Configurable region file format
 
             this.regionCache.putAndMoveToFirst(key, ret);
 
@@ -312,7 +330,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
         LOGGER.error("{} ({} - {},{}) Go clean it up to remove this message. /minecraft:tp {} 128 {} - DO NOT REPORT THIS TO {} - You may ask for help on Discord, but do not file an issue. These error messages can not be removed.", msg, file.toString().replaceAll(".+[\\\\/]", ""), x, z, x << 4, z << 4, io.papermc.paper.ServerBuildInfo.buildInfo().brandName().toUpperCase()); // Canvas - Rebrand
     }
 
-    private static CompoundTag readOversizedChunk(RegionFile regionfile, ChunkPos chunkCoordinate) throws IOException {
+    private static CompoundTag readOversizedChunk(io.canvasmc.canvas.regionformat.IRegionFile regionfile, ChunkPos chunkCoordinate) throws IOException { // Linear - Configurable region file format
         synchronized (regionfile) {
             try (DataInputStream datainputstream = regionfile.getChunkDataInputStream(chunkCoordinate)) {
                 CompoundTag oversizedData = regionfile.getOversizedData(chunkCoordinate.x(), chunkCoordinate.z());
@@ -346,7 +364,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
     public @Nullable CompoundTag read(final ChunkPos pos) throws IOException {
         // CraftBukkit start - SPIGOT-5680: There's no good reason to preemptively create files on read, save that for writing
-        RegionFile region = this.getRegionFile(pos, true);
+        io.canvasmc.canvas.regionformat.IRegionFile region = this.getRegionFile(pos, true); // Luminol - Configurable region file format
         if (region == null) {
             return null;
         }
@@ -383,7 +401,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
     public void scanChunk(final ChunkPos pos, final StreamTagVisitor scanner) throws IOException {
         // CraftBukkit start - SPIGOT-5680: There's no good reason to preemptively create files on read, save that for writing
-        RegionFile region = this.getRegionFile(pos, true);
+        io.canvasmc.canvas.regionformat.IRegionFile region = this.getRegionFile(pos, true); // Luminol - Configurable region file format
         if (region == null) {
             return;
         }
@@ -398,7 +416,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
     public void write(final ChunkPos pos, final @Nullable CompoundTag value) throws IOException {
         if (!SharedConstants.DEBUG_DONT_SAVE_WORLD) {
-            RegionFile region = this.getRegionFile(pos, value == null); // CraftBukkit // Paper - rewrite chunk system
+            io.canvasmc.canvas.regionformat.IRegionFile region = this.getRegionFile(pos, value == null); // CraftBukkit // Paper - rewrite chunk system // Luminol - Configurable region file format
             // Paper start - rewrite chunk system
             if (region == null) {
                 // if the RegionFile doesn't exist, no point in deleting from it
@@ -430,7 +448,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
         // Paper start - rewrite chunk system
         synchronized (this) {
             final ExceptionCollector<IOException> exceptionCollector = new ExceptionCollector<>();
-            for (final RegionFile regionFile : this.regionCache.values()) {
+            for (final io.canvasmc.canvas.regionformat.IRegionFile regionFile : this.regionCache.values()) { // Luminol - Configurable region file format
                 try {
                     regionFile.close();
                 } catch (final IOException ex) {
@@ -446,7 +464,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
         // Paper start - rewrite chunk system
         synchronized (this) {
             final ExceptionCollector<IOException> exceptionCollector = new ExceptionCollector<>();
-            for (final RegionFile regionFile : this.regionCache.values()) {
+            for (final io.canvasmc.canvas.regionformat.IRegionFile regionFile : this.regionCache.values()) { // Luminol - Configurable region file format
                 try {
                     regionFile.flush();
                 } catch (final IOException ex) {
